File: frontend/src/types/whoop.ts
/**
 * TypeScript types for Whoop API integration.
 */

// =============================================================================
// OAuth Types
// =============================================================================

export interface WhoopConnectResponse {
  authorization_url: string;
  state: string;
}

export interface WhoopConnectionStatus {
  is_connected: boolean;
  whoop_user_id: string | null;
  connected_at: string | null;
  last_sync_at: string | null;
  scopes: string[];
}

export interface WhoopDisconnectResponse {
  success: boolean;
  message: string;
}

// =============================================================================
// Sync Types
// =============================================================================

export interface WhoopSyncRequest {
  start_date?: string;
  end_date?: string;
}

export interface WhoopSyncResponse {
  success: boolean;
  cycles_synced: number;
  recovery_synced: number;
  sleep_synced: number;
  workouts_synced: number;
  sync_completed_at: string;
}

// =============================================================================
// Data Types
// =============================================================================

export interface WhoopCycle {
  id: string;
  whoop_cycle_id: string; // UUID in v2 API
  start_time: string;
  end_time: string | null;
  strain_score: number | null;
  kilojoules: number | null;
  average_heart_rate: number | null;
  max_heart_rate: number | null;
  created_at: string;
}

export interface WhoopRecovery {
  id: string;
  whoop_cycle_id: string; // UUID in v2 API
  recovery_score: number | null;
  resting_heart_rate: number | null;
  hrv_rmssd_milli: number | null;
  spo2_percentage: number | null;
  skin_temp_celsius: number | null;
  created_at: string;
}

export interface WhoopSleep {
  id: string;
  whoop_sleep_id: string; // UUID in v2 API
  whoop_cycle_id: string | null; // UUID in v2 API
  start_time: string;
  end_time: string;
  is_nap: boolean;
  sleep_score: number | null;
  total_in_bed_milli: number | null;
  total_awake_milli: number | null;
  total_light_sleep_milli: number | null;
  total_slow_wave_sleep_milli: number | null;
  total_rem_sleep_milli: number | null;
  sleep_efficiency: number | null;
  respiratory_rate: number | null;
  created_at: string;
}

export interface WhoopWorkout {
  id: string;
  whoop_workout_id: string; // UUID in v2 API
  whoop_cycle_id: string | null; // UUID in v2 API
  start_time: string;
  end_time: string;
  sport_id: number;
  sport_name: string | null;
  strain_score: number | null;
  kilojoules: number | null;
  average_heart_rate: number | null;
  max_heart_rate: number | null;
  distance_meter: number | null;
  altitude_gain_meter: number | null;
  created_at: string;
}

// =============================================================================
// Dashboard Types
// =============================================================================

export interface WhoopDashboardSummary {
  is_connected: boolean;
  last_sync_at: string | null;
  latest_recovery_score: number | null;
  latest_strain_score: number | null;
  latest_hrv: number | null;
  latest_resting_hr: number | null;
  latest_sleep_score: number | null;
  latest_sleep_hours: number | null;
  avg_recovery_7d: number | null;
  avg_strain_7d: number | null;
  avg_sleep_hours_7d: number | null;
  total_workouts_7d: number;
}

// =============================================================================
// List Response Types
// =============================================================================

export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  page: number;
  page_size: number;
}

export type WhoopSleepListResponse = PaginatedResponse<WhoopSleep>;
export type WhoopWorkoutListResponse = PaginatedResponse<WhoopWorkout>;
export type WhoopRecoveryListResponse = PaginatedResponse<WhoopRecovery>;
export type WhoopCycleListResponse = PaginatedResponse<WhoopCycle>;

// =============================================================================
// Helper Types
// =============================================================================

export type WhoopDataType = 'sleep' | 'workouts' | 'recovery' | 'cycles';

export interface WhoopContextState {
  connectionStatus: WhoopConnectionStatus | null;
  dashboardSummary: WhoopDashboardSummary | null;
  isLoading: boolean;
  isSyncing: boolean;
  error: string | null;
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Convert milliseconds to hours with 1 decimal place.
 */
export function millisToHours(milli: number | null): number | null {
  if (milli === null) return null;
  return Math.round((milli / 3600000) * 10) / 10;
}

/**
 * Format a recovery score with color indication.
 */
export function getRecoveryColor(score: number | null): string {
  if (score === null) return 'gray';
  if (score >= 67) return 'green';
  if (score >= 34) return 'yellow';
  return 'red';
}

/**
 * Format strain score with color indication.
 */
export function getStrainColor(score: number | null): string {
  if (score === null) return 'gray';
  if (score >= 18) return 'red';
  if (score >= 14) return 'orange';
  if (score >= 10) return 'yellow';
  return 'blue';
}

/**
 * Get human-readable sport name from sport_id.
 */
export function getSportName(sportId: number, sportName: string | null): string {
  if (sportName) return sportName;

  // Common Whoop sport IDs
  const sportMap: Record<number, string> = {
    0: 'Activity',
    1: 'Running',
    33: 'Cycling',
    44: 'Swimming',
    48: 'Yoga',
    52: 'Weightlifting',
    71: 'CrossFit',
    82: 'HIIT',
    // Add more as needed
  };

  return sportMap[sportId] || `Activity (${sportId})`;
}

/**
 * Format date for display.
 */
export function formatWhoopDate(dateString: string): string {
  const date = new Date(dateString);
  return date.toLocaleDateString('en-US', {
    month: 'short',
    day: 'numeric',
    year: 'numeric',
  });
}

/**
 * Format time for display.
 */
export function formatWhoopTime(dateString: string): string {
  const date = new Date(dateString);
  return date.toLocaleTimeString('en-US', {
    hour: 'numeric',
    minute: '2-digit',
  });
}

File: backend/app/schemas/whoop.py
"""
Pydantic schemas for Whoop API integration.

Defines request/response models for:
- OAuth connection flow
- Data sync operations
- Sleep, recovery, workout, and cycle data
"""

from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


# =============================================================================
# OAuth Schemas
# =============================================================================


class WhoopConnectResponse(BaseModel):
    """Response with OAuth authorization URL."""

    authorization_url: str = Field(..., description="URL to redirect user for Whoop authorization")
    state: str = Field(..., description="CSRF state parameter")


class WhoopCallbackRequest(BaseModel):
    """OAuth callback parameters."""

    code: str = Field(..., description="Authorization code from Whoop")
    state: str = Field(..., description="State parameter for CSRF verification")


class WhoopConnectionStatus(BaseModel):
    """Current Whoop connection status."""

    is_connected: bool = Field(..., description="Whether Whoop is connected")
    whoop_user_id: Optional[str] = Field(None, description="Whoop user ID if connected")
    connected_at: Optional[datetime] = Field(None, description="When connection was established")
    last_sync_at: Optional[datetime] = Field(None, description="Last successful sync time")
    scopes: list[str] = Field(default_factory=list, description="Granted OAuth scopes")


class WhoopDisconnectResponse(BaseModel):
    """Response after disconnecting Whoop."""

    success: bool = Field(..., description="Whether disconnect was successful")
    message: str = Field(..., description="Status message")


# =============================================================================
# Sync Schemas
# =============================================================================


class WhoopSyncRequest(BaseModel):
    """Request to sync Whoop data."""

    start_date: Optional[datetime] = Field(None, description="Start date for sync range")
    end_date: Optional[datetime] = Field(None, description="End date for sync range")


class WhoopSyncResponse(BaseModel):
    """Response after syncing Whoop data."""

    success: bool = Field(..., description="Whether sync was successful")
    cycles_synced: int = Field(0, description="Number of cycles synced")
    recovery_synced: int = Field(0, description="Number of recovery records synced")
    sleep_synced: int = Field(0, description="Number of sleep records synced")
    workouts_synced: int = Field(0, description="Number of workouts synced")
    sync_completed_at: datetime = Field(..., description="When sync completed")


# =============================================================================
# Cycle Schemas
# =============================================================================


class WhoopCycleBase(BaseModel):
    """Base schema for cycle data."""

    whoop_cycle_id: str = Field(..., description="Whoop's cycle ID (UUID in v2)")
    start_time: datetime = Field(..., description="Cycle start time")
    end_time: Optional[datetime] = Field(None, description="Cycle end time")
    strain_score: Optional[Decimal] = Field(None, ge=0, le=21, description="Day strain (0-21)")
    kilojoules: Optional[Decimal] = Field(None, ge=0, description="Calories burned in kJ")
    average_heart_rate: Optional[int] = Field(None, ge=0, description="Average heart rate")
    max_heart_rate: Optional[int] = Field(None, ge=0, description="Max heart rate")


class WhoopCycleResponse(WhoopCycleBase):
    """Cycle data response."""

    id: str = Field(..., description="Database UUID")
    created_at: datetime = Field(..., description="Record creation time")

    class Config:
        from_attributes = True


# =============================================================================
# Recovery Schemas
# =============================================================================


class WhoopRecoveryBase(BaseModel):
    """Base schema for recovery data."""

    whoop_cycle_id: str = Field(..., description="Associated cycle ID (UUID in v2)")
    recovery_score: Optional[Decimal] = Field(None, ge=0, le=100, description="Recovery score (0-100%)")
    resting_heart_rate: Optional[Decimal] = Field(None, ge=0, description="Resting heart rate")
    hrv_rmssd_milli: Optional[Decimal] = Field(None, ge=0, description="HRV in milliseconds")
    spo2_percentage: Optional[Decimal] = Field(None, ge=0, le=100, description="Blood oxygen %")
    skin_temp_celsius: Optional[Decimal] = Field(None, description="Skin temperature in Celsius")


class WhoopRecoveryResponse(WhoopRecoveryBase):
    """Recovery data response."""

    id: str = Field(..., description="Database UUID")
    created_at: datetime = Field(..., description="Record creation time")

    class Config:
        from_attributes = True


# =============================================================================
# Sleep Schemas
# =============================================================================


class WhoopSleepBase(BaseModel):
    """Base schema for sleep data."""

    whoop_sleep_id: str = Field(..., description="Whoop's sleep ID (UUID in v2)")
    whoop_cycle_id: Optional[str] = Field(None, description="Associated cycle ID (UUID in v2)")
    start_time: datetime = Field(..., description="Sleep start time")
    end_time: datetime = Field(..., description="Sleep end time")
    is_nap: bool = Field(False, description="Whether this is a nap")
    sleep_score: Optional[Decimal] = Field(None, ge=0, le=100, description="Sleep performance score")
    total_in_bed_milli: Optional[int] = Field(None, ge=0, description="Total time in bed (ms)")
    total_awake_milli: Optional[int] = Field(None, ge=0, description="Time awake (ms)")
    total_light_sleep_milli: Optional[int] = Field(None, ge=0, description="Light sleep time (ms)")
    total_slow_wave_sleep_milli: Optional[int] = Field(None, ge=0, description="Deep sleep time (ms)")
    total_rem_sleep_milli: Optional[int] = Field(None, ge=0, description="REM sleep time (ms)")
    sleep_efficiency: Optional[Decimal] = Field(None, ge=0, le=1, description="Sleep efficiency (0-1)")
    respiratory_rate: Optional[Decimal] = Field(None, ge=0, description="Respiratory rate")


class WhoopSleepResponse(WhoopSleepBase):
    """Sleep data response."""

    id: str = Field(..., description="Database UUID")
    created_at: datetime = Field(..., description="Record creation time")

    # Computed fields for convenience
    @property
    def total_sleep_hours(self) -> Optional[float]:
        """Calculate total sleep time in hours."""
        if self.total_in_bed_milli and self.total_awake_milli:
            sleep_milli = self.total_in_bed_milli - self.total_awake_milli
            return round(sleep_milli / 3600000, 2)
        return None

    class Config:
        from_attributes = True


# =============================================================================
# Workout Schemas
# =============================================================================


class WhoopWorkoutBase(BaseModel):
    """Base schema for workout data."""

    whoop_workout_id: str = Field(..., description="Whoop's workout ID (UUID in v2)")
    whoop_cycle_id: Optional[str] = Field(None, description="Associated cycle ID (UUID in v2)")
    start_time: datetime = Field(..., description="Workout start time")
    end_time: datetime = Field(..., description="Workout end time")
    sport_id: int = Field(..., description="Sport type ID")
    sport_name: Optional[str] = Field(None, description="Sport name")
    strain_score: Optional[Decimal] = Field(None, ge=0, le=21, description="Workout strain")
    kilojoules: Optional[Decimal] = Field(None, ge=0, description="Calories burned in kJ")
    average_heart_rate: Optional[int] = Field(None, ge=0, description="Average heart rate")
    max_heart_rate: Optional[int] = Field(None, ge=0, description="Max heart rate")
    distance_meter: Optional[Decimal] = Field(None, ge=0, description="Distance in meters")
    altitude_gain_meter: Optional[Decimal] = Field(None, description="Altitude gained in meters")


class WhoopWorkoutResponse(WhoopWorkoutBase):
    """Workout data response."""

    id: str = Field(..., description="Database UUID")
    created_at: datetime = Field(..., description="Record creation time")

    @property
    def duration_minutes(self) -> float:
        """Calculate workout duration in minutes."""
        delta = self.end_time - self.start_time
        return round(delta.total_seconds() / 60, 1)

    class Config:
        from_attributes = True


# =============================================================================
# Dashboard Summary Schemas
# =============================================================================


class WhoopDashboardSummary(BaseModel):
    """Summary data for dashboard display."""

    # Connection status
    is_connected: bool = Field(..., description="Whether Whoop is connected")
    last_sync_at: Optional[datetime] = Field(None, description="Last sync time")

    # Latest metrics (from most recent cycle)
    latest_recovery_score: Optional[Decimal] = Field(None, description="Most recent recovery %")
    latest_strain_score: Optional[Decimal] = Field(None, description="Most recent strain")
    latest_hrv: Optional[Decimal] = Field(None, description="Most recent HRV")
    latest_resting_hr: Optional[Decimal] = Field(None, description="Most recent resting HR")

    # Latest sleep (from most recent sleep record)
    latest_sleep_score: Optional[Decimal] = Field(None, description="Most recent sleep score")
    latest_sleep_hours: Optional[float] = Field(None, description="Most recent sleep duration")

    # Aggregates (last 7 days)
    avg_recovery_7d: Optional[Decimal] = Field(None, description="7-day average recovery")
    avg_strain_7d: Optional[Decimal] = Field(None, description="7-day average strain")
    avg_sleep_hours_7d: Optional[float] = Field(None, description="7-day average sleep hours")
    total_workouts_7d: int = Field(0, description="Workouts in last 7 days")


# =============================================================================
# List Response Schemas
# =============================================================================


class WhoopCycleListResponse(BaseModel):
    """Paginated list of cycles."""

    data: list[WhoopCycleResponse] = Field(default_factory=list)
    total: int = Field(0, description="Total records available")
    page: int = Field(1, description="Current page")
    page_size: int = Field(10, description="Records per page")


class WhoopRecoveryListResponse(BaseModel):
    """Paginated list of recovery records."""

    data: list[WhoopRecoveryResponse] = Field(default_factory=list)
    total: int = Field(0, description="Total records available")
    page: int = Field(1, description="Current page")
    page_size: int = Field(10, description="Records per page")


class WhoopSleepListResponse(BaseModel):
    """Paginated list of sleep records."""

    data: list[WhoopSleepResponse] = Field(default_factory=list)
    total: int = Field(0, description="Total records available")
    page: int = Field(1, description="Current page")
    page_size: int = Field(10, description="Records per page")


class WhoopWorkoutListResponse(BaseModel):
    """Paginated list of workouts."""

    data: list[WhoopWorkoutResponse] = Field(default_factory=list)
    total: int = Field(0, description="Total records available")
    page: int = Field(1, description="Current page")
    page_size: int = Field(10, description="Records per page")
